### ISSUE TITLE
Unable to install on MacOS

### ISSUE THREAD

**Issue by jonathanabennett** [lem-project/lem#573]
I installed Roswell on my M1 Mac via Homebrew, then ran `ros install lem-project/lem`. When I do that, I receive the following error:

Installing from github lem-project/lem
To load "lem":
  Load 1 ASDF system:
    lem
; Loading "lem"
..............[1/3] System 'lem' found. Loading the system..Aborted during step [1/3].
Unhandled CFFI:LOAD-FOREIGN-LIBRARY-ERROR in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                                       {7005E4A903}>:
  Unable to load foreign library (ASYNC-PROCESS).
  Error opening shared object "/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/static/x86_64/Darwin/libasyncprocess.so":
  dlopen(/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/static/x86_64/Darwin/libasyncprocess.so, 0x000A): tried: '/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/static/x86_64/Darwin/libasyncprocess.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')).

Backtrace for: #<SB-THREAD:THREAD "main thread" RUNNING {7005E4A903}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<CFFI:LOAD-FOREIGN-LIBRARY-ERROR "Unable to load foreign library (~A).~%  ~A" {7008A21213}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK SB-EXT:*INVOKE-DEBUGGER-HOOK* #<CFFI:LOAD-FOREIGN-LIBRARY-ERROR "Unable to load foreign library (~A).~%  ~A" {7008A21213}>)
2: (INVOKE-DEBUGGER #<CFFI:LOAD-FOREIGN-LIBRARY-ERROR "Unable to load foreign library (~A).~%  ~A" {7008A21213}>)
3: (ERROR CFFI:LOAD-FOREIGN-LIBRARY-ERROR :FORMAT-CONTROL "Unable to load foreign library (~A).~%  ~A" :FORMAT-ARGUMENTS (ASYNC-PROCESS #<(SIMPLE-BASE-STRING 561) Error opening shared object "/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/static/x86_64/Darwin/libasyncprocess.so":
  dlopen(/Users/jonathan.benn... {7008A20FCF}>))
4: (CFFI::FL-ERROR "Unable to load foreign library (~A).~%  ~A" ASYNC-PROCESS #<(SIMPLE-BASE-STRING 561) Error opening shared object "/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/static/x86_64/Darwin/libasyncprocess.so":
  dlopen(/Users/jonathan.benn... {7008A20FCF}>)
5: (CFFI::REPORT-SIMPLE-ERROR ASYNC-PROCESS #<SIMPLE-ERROR "Error opening ~:[runtime~;shared object ~:*~S~]:
  ~A." {7008A0F653}>)
6: ((FLET CFFI::%DO-LOAD :IN CFFI::%DO-LOAD-FOREIGN-LIBRARY) #<CFFI:FOREIGN-LIBRARY ASYNC-PROCESS> ASYNC-PROCESS "libasyncprocess.so")
7: (CFFI:LOAD-FOREIGN-LIBRARY ASYNC-PROCESS :SEARCH-PATH NIL)
8: (SB-FASL::LOAD-FASL-GROUP #S(SB-FASL::FASL-INPUT :STREAM #<SB-SYS:FD-STREAM for "file /Users/jonathan.bennett/.cache/common-lisp/sbcl-2.2.2-macosx-arm64/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/src/async-process.fasl" {70089E5493}> :TABLE #(191 #<PACKAGE "SB-IMPL"> SB-IMPL::%DEFPACKAGE "ASYNC-PROCESS" #1="CL" (#1#) #2="DELETE-PROCESS" #3="PROCESS-SEND-INPUT" #4="PROCESS-RECEIVE-OUTPUT" #5="PROCESS-ALIVE-P" #6="CREATE-PROCESS" (#2# #3# #4# #5# #6#) ...) :STACK #(0 CFFI:LOAD-FOREIGN-LIBRARY ASYNC-PROCESS ((:UNIX "libasyncprocess.so") #7=(:WINDOWS #8="libasyncprocess.dll")) #7# #8# NIL (SB-INT:SFUNCTION NIL NULL) #<SB-KERNEL:FDEFN UIOP/OS:FEATUREP> #<SB-KERNEL:FDEFN SB-IMPL::%INIT-STRING-OUTPUT-STREAM> #<SB-KERNEL:FDEFN WRITE-STRING> #<SB-KERNEL:FDEFN PRINC> ...) :NAME-BUFFER #("FEATUREP-EQUALE-PATHNAME" "LOAD-FOREIGN-LIBRARYRARYAMES*") :DEPRECATED-STUFF NIL :SKIP-UNTIL NIL) NIL)
9: (SB-FASL::LOAD-AS-FASL #<SB-SYS:FD-STREAM for "file /Users/jonathan.bennett/.cache/common-lisp/sbcl-2.2.2-macosx-arm64/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/src/async-process.fasl" {70089E5493}> NIL NIL)
10: ((LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD) #<SB-SYS:FD-STREAM for "file /Users/jonathan.bennett/.cache/common-lisp/sbcl-2.2.2-macosx-arm64/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/src/async-process.fasl" {70089E5493}> T)
11: (SB-FASL::CALL-WITH-LOAD-BINDINGS #<FUNCTION (LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD) {104522AAB}> #<SB-SYS:FD-STREAM for "file /Users/jonathan.bennett/.cache/common-lisp/sbcl-2.2.2-macosx-arm64/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/src/async-process.fasl" {70089E5493}> T #<SB-SYS:FD-STREAM for "file /Users/jonathan.bennett/.cache/common-lisp/sbcl-2.2.2-macosx-arm64/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/src/async-process.fasl" {70089E5493}>)
12: (LOAD #P"/Users/jonathan.bennett/.cache/common-lisp/sbcl-2.2.2-macosx-arm64/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/src/async-process.fasl" :VERBOSE NIL :PRINT NIL :IF-DOES-NOT-EXIST T :EXTERNAL-FORMAT :DEFAULT)
13: (UIOP/UTILITY:CALL-WITH-MUFFLED-CONDITIONS #<FUNCTION (LAMBDA NIL :IN UIOP/LISP-BUILD:LOAD*) {70089E4B5B}> ("Overwriting already existing readtable ~S." #(#:FINALIZERS-OFF-WARNING :ASDF-FINALIZERS)))
14: ((SB-PCL::EMF ASDF/ACTION:PERFORM) #<unused argument> #<unused argument> #<ASDF/LISP-ACTION:LOAD-OP > #<ASDF/LISP-ACTION:CL-SOURCE-FILE "async-process" "async-process">)
15: ((LAMBDA NIL :IN ASDF/ACTION:CALL-WHILE-VISITING-ACTION))
16: ((:METHOD ASDF/ACTION:PERFORM-WITH-RESTARTS (ASDF/LISP-ACTION:LOAD-OP ASDF/LISP-ACTION:CL-SOURCE-FILE)) #<ASDF/LISP-ACTION:LOAD-OP > #<ASDF/LISP-ACTION:CL-SOURCE-FILE "async-process" "async-process">) [fast-method]
17: ((:METHOD ASDF/ACTION:PERFORM-WITH-RESTARTS :AROUND (T T)) #<ASDF/LISP-ACTION:LOAD-OP > #<ASDF/LISP-ACTION:CL-SOURCE-FILE "async-process" "async-process">) [fast-method]
18: ((:METHOD ASDF/PLAN:PERFORM-PLAN (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {70052A2F93}>) [fast-method]
19: ((FLET SB-C::WITH-IT :IN SB-C::%WITH-COMPILATION-UNIT))
20: ((:METHOD ASDF/PLAN:PERFORM-PLAN :AROUND (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {70052A2F93}>) [fast-method]
21: ((:METHOD ASDF/OPERATE:OPERATE (ASDF/OPERATION:OPERATION ASDF/COMPONENT:COMPONENT)) #<ASDF/LISP-ACTION:LOAD-OP > #<ASDF/SYSTEM:SYSTEM "lem"> :PLAN-CLASS NIL :PLAN-OPTIONS NIL) [fast-method]
22: ((SB-PCL::EMF ASDF/OPERATE:OPERATE) #<unused argument> #<unused argument> #<ASDF/LISP-ACTION:LOAD-OP > #<ASDF/SYSTEM:SYSTEM "lem"> :VERBOSE NIL)
23: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
24: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) #<ASDF/LISP-ACTION:LOAD-OP > #<ASDF/SYSTEM:SYSTEM "lem"> :VERBOSE NIL) [fast-method]
25: ((SB-PCL::EMF ASDF/OPERATE:OPERATE) #<unused argument> #<unused argument> ASDF/LISP-ACTION:LOAD-OP "lem" :VERBOSE NIL)
26: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
27: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) ASDF/LISP-ACTION:LOAD-OP "lem" :VERBOSE NIL) [fast-method]
28: (ASDF/SESSION:CALL-WITH-ASDF-SESSION #<FUNCTION (LAMBDA NIL :IN ASDF/OPERATE:OPERATE) {70052A2D8B}> :OVERRIDE T :KEY NIL :OVERRIDE-CACHE T :OVERRIDE-FORCING NIL)
29: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
30: (ASDF/SESSION:CALL-WITH-ASDF-SESSION #<FUNCTION (LAMBDA NIL :IN ASDF/OPERATE:OPERATE) {70052A2C7B}> :OVERRIDE NIL :KEY NIL :OVERRIDE-CACHE NIL :OVERRIDE-FORCING NIL)
31: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) ASDF/LISP-ACTION:LOAD-OP "lem" :VERBOSE NIL) [fast-method]
32: (ASDF/OPERATE:LOAD-SYSTEM "lem" :VERBOSE NIL)
33: (QUICKLISP-CLIENT::CALL-WITH-MACROEXPAND-PROGRESS #<FUNCTION (LAMBDA NIL :IN QUICKLISP-CLIENT::APPLY-LOAD-STRATEGY) {70052A2AEB}>)
34: (QUICKLISP-CLIENT::AUTOLOAD-SYSTEM-AND-DEPENDENCIES "lem" :PROMPT NIL)
35: ((:METHOD QL-IMPL-UTIL::%CALL-WITH-QUIET-COMPILATION (T T)) #<unused argument> #<FUNCTION (FLET QUICKLISP-CLIENT::QL :IN QUICKLISP-CLIENT:QUICKLOAD) {70052A298B}>) [fast-method]
36: ((:METHOD QL-IMPL-UTIL::%CALL-WITH-QUIET-COMPILATION :AROUND (QL-IMPL:SBCL T)) #<QL-IMPL:SBCL {70052A29F3}> #<FUNCTION (FLET QUICKLISP-CLIENT::QL :IN QUICKLISP-CLIENT:QUICKLOAD) {70052A298B}>) [fast-method]
37: ((:METHOD QUICKLISP-CLIENT:QUICKLOAD (T)) "lem" :PROMPT NIL :SILENT NIL :VERBOSE NIL) [fast-method]
38: (QL-DIST::CALL-WITH-CONSISTENT-DISTS #<FUNCTION (LAMBDA NIL :IN QUICKLISP-CLIENT:QUICKLOAD) {70052A295B}>)
39: (ROSWELL.INSTALL:INSTALL-SYSTEM-SCRIPT "lem")
40: (ROSWELL.INSTALL:INSTALL-SYSTEM-IF-PROBED "lem")
41: (ROSWELL.INSTALL:INSTALL NIL)
42: (SB-INT:SIMPLE-EVAL-IN-LEXENV (APPLY (QUOTE ROS.SCRIPT.INSTALL.3671936092::MAIN) ROSWELL:*ARGV*) #<NULL-LEXENV>)
43: (SB-INT:SIMPLE-EVAL-IN-LEXENV (ROSWELL:QUIT (APPLY (QUOTE ROS.SCRIPT.INSTALL.3671936092::MAIN) ROSWELL:*ARGV*)) #<NULL-LEXENV>)
44: (SB-EXT:EVAL-TLF (ROSWELL:QUIT (APPLY (QUOTE ROS.SCRIPT.INSTALL.3671936092::MAIN) ROSWELL:*ARGV*)) NIL NIL)
45: ((LABELS SB-FASL::EVAL-FORM :IN SB-INT:LOAD-AS-SOURCE) (ROSWELL:QUIT (APPLY (QUOTE ROS.SCRIPT.INSTALL.3671936092::MAIN) ROSWELL:*ARGV*)) NIL)
46: (SB-INT:LOAD-AS-SOURCE #<CONCATENATED-STREAM :STREAMS NIL {70052A2533}> :VERBOSE NIL :PRINT NIL :CONTEXT "loading")
47: ((LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD) #<CONCATENATED-STREAM :STREAMS NIL {70052A2533}> NIL)
48: (SB-FASL::CALL-WITH-LOAD-BINDINGS #<FUNCTION (LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD) {104520A7B}> #<CONCATENATED-STREAM :STREAMS NIL {70052A2533}> NIL #<CONCATENATED-STREAM :STREAMS NIL {70052A2533}>)
49: (LOAD #<CONCATENATED-STREAM :STREAMS NIL {70052A2533}> :VERBOSE NIL :PRINT NIL :IF-DOES-NOT-EXIST T :EXTERNAL-FORMAT :DEFAULT)
50: ((FLET ROSWELL::BODY :IN ROSWELL:SCRIPT) #<SB-SYS:FD-STREAM for "file /opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros" {70052A2183}>)
51: (ROSWELL:SCRIPT "/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros" "lem-project/lem")
52: (ROSWELL:RUN ((:EVAL "(ros:quicklisp)") (:SCRIPT "/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros" "lem-project/lem") (:QUIT NIL)))
53: (SB-INT:SIMPLE-EVAL-IN-LEXENV (ROSWELL:RUN (QUOTE ((:EVAL "(ros:quicklisp)") (:SCRIPT "/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros" "lem-project/lem") (:QUIT NIL)))) #<NULL-LEXENV>)
54: (EVAL (ROSWELL:RUN (QUOTE ((:EVAL "(ros:quicklisp)") (:SCRIPT "/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros" "lem-project/lem") (:QUIT NIL)))))
55: (SB-IMPL::PROCESS-EVAL/LOAD-OPTIONS ((:EVAL . "(progn #-ros.init(cl:load \"/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/init.lisp\"))") (:EVAL . "(ros:run '((:eval\"(ros:quicklisp)\")(:script \"/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros\"\"lem-project/lem\")(:quit ())))")))
56: (SB-IMPL::TOPLEVEL-INIT)
57: ((FLET SB-UNIX::BODY :IN SB-IMPL::START-LISP))
58: ((FLET "WITHOUT-INTERRUPTS-BODY-3" :IN SB-IMPL::START-LISP))
59: (SB-IMPL::START-LISP)

unhandled condition in --disable-debugger mode, quitting
;
; compilation unit aborted
;   caught 1 fatal ERROR condition

**Comment by Angel-Cortez** [comment]
@jonathanabennett, I'm not sure if this project is M1 compatible yet

**Comment by snmsts** [comment]
I'm using lem on m1 and it works fine for me.
I guess ``uname -m`` from sbcl process outputs ``x86_64``

Would you share result of 

```
ros -v install lem-project/lem
```

?

**Comment by jonathanabennett** [comment]
Here is the result:

opt:verbose:1
dispatch:install,name=top
proc_cmd:
cmd_install:
get_opt(variant,0)=(null)
/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros
lem-project has not been implemented internally yet. /opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros argc:2
install:lem-project/lem:
argc_=3 argv[0]=--,argv[1]=/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros,argv[2]=lem-project/lem,dispatch:--,name=top
proc_options:--:top
cmd_script_frontend:0
frontend:script_:argc=2 argv[0]=/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros
ros_script_cmd=exec ros -Q +R -L sbcl-bin -- $0 "$@"
dispatch:-Q,name=top
proc_options:-Q:top
opt:quicklisp:2
dispatch:+R,name=top
proc_options:+R:top
opt:no-rc:0
dispatch:-L,name=top
proc_options:-L:top
get_opt(lisp,0)=(null)
take1:lisp:sbcl-bin,(null)
set_opt(lisp)='sbcl-bin'
dispatch:--,name=top
proc_options:--:top
cmd_script_frontend:1
cmd_script
get_opt(program,0)=(null)
script_:argc=3 argv[0]=--
current=(null)
set_opt(script)='"/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros""lem-project/lem"'
cmd_run_star:1:argv[0],script
get_opt(image,0)=(null)
get_opt(roswellenv,0)=(null)
set_env_opt:.roswellenv
get_opt(roswellenv,1)=(null)
set_env_opt:/Users/jonathan.bennett/.roswell/env/
get_opt(lisp,1)=sbcl-bin
determin_impl:sbcl-bin
get_opt(sbcl-bin.version,1)=2.2.2
set_opt(impl)='sbcl-bin/2.2.2'
get_opt(quicklisp,0)=(null)
set_opt(quicklisp)='/Users/jonathan.bennett/.roswell/lisp/quicklisp/'
set_opt(argv0)='ros'
which cmd:command -v "ros"
which result:/opt/homebrew/bin/ros

set_opt(wargv0)='/opt/homebrew/bin/ros'
set_opt(homedir)='/Users/jonathan.bennett/.roswell/'
set_opt(verbose)='1'
set_opt(lispdir)='/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/'
set_opt(patchdir)='/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/patch/'
get_opt(asdf.version,0)=(null)
set_opt(uname)='darwin'
set_opt(uname-m)='arm64'
get_opt(program,0)=(null)
get_opt(asdf.version,0)=(null)
set_opt(program)='(:eval"(ros:quicklisp)")'
get_opt(impl,0)=sbcl-bin/2.2.2
get_opt(help,0)=(null)
get_opt(script,0)="/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros""lem-project/lem"
get_opt(image,0)=(null)
get_opt(program,0)=(:eval"(ros:quicklisp)")
get_opt(dynamic-space-size,1)=(null)
get_opt(control-stack-size,1)=(null)
get_opt(without-roswell,0)=(null)
get_opt(enable-debugger,0)=(null)
get_opt(version,0)=(null)
get_opt(asdf.version,0)=(null)
init.lisp=/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/init.lisp

help=nil script="/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros""lem-project/lem"
get_opt(wrap,1)=(null)
args /Users/jonathan.bennett/.roswell/impls/arm64/darwin/sbcl-bin/2.2.2/bin/sbcl --core /Users/jonathan.bennett/.roswell/impls/arm64/darwin/sbcl-bin/2.2.2/lib/sbcl/sbcl.core --noinform --no-sysinit --no-userinit --disable-debugger --eval (progn #-ros.init(cl:load "/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/init.lisp")) --eval (ros:run '((:eval"(ros:quicklisp)")(:script "/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros""lem-project/lem")(:quit ())))
ROS_OPTS (("program""(:eval\"(ros:quicklisp)\")")("uname-m""arm64")("uname""darwin")("patchdir""/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/patch/")("lispdir""/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/")("verbose""1")("homedir""/Users/jonathan.bennett/.roswell/")("wargv0""/opt/homebrew/bin/ros")("argv0""ros")("quicklisp""/Users/jonathan.bennett/.roswell/lisp/quicklisp/")("impl""sbcl-bin/2.2.2")("script""\"/opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/install.ros\"\"lem-project/lem\"")("lisp""sbcl-bin")("setup.time""3859849472")("sbcl-bin.version""2.2.2")("default.lisp""sbcl-bin"))
; loading #P"/Users/jonathan.bennett/.roswell/lisp/quicklisp/setup.lisp"
LOAD /opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/patch-quicklisp.lisp
LOAD /opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/util.lisp
LOAD /opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/extend-quicklisp-system.lisp
LOAD /opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/util-install.lisp
LOAD /opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/locations.lisp
LOAD /opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/util-install-quicklisp.lisp
LOAD /opt/homebrew/Cellar/roswell/21.10.14.111/etc/roswell/system.lisp
Installing from github lem-project/lem
[1/3] System 'lem' found. Loading the system..%

It exits at the % sign

**Comment by snmsts** [comment]
Thank you for feedback.

mmm... it seems sbcl loaded is the one for arm64... that's expected behavior. 
I'm author of both roswell and async-process's code to determine which arch are used there.

Let me confirm from basic. I believe on your environment

```
ros roswell-internal-use uname -m
```

should be 

```
arm64
```

and still failing with

```
ros -s async-process
```

by trying to load x86_64's so?

**Comment by jonathanabennett** [comment]
That is correct. Those commands both produce the output you predicted. How do I get roswell to download the sbcl for arm64?

**Comment by jonathanabennett** [comment]
If it helps, here's the exact output of running `ros -s async-process`

`ros -s async-process

debugger invoked on a CFFI:LOAD-FOREIGN-LIBRARY-ERROR in thread
#<THREAD "main thread" RUNNING {7008650203}>:
  Unable to load foreign library (ASYNC-PROCESS).
  Error opening shared object "/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/static/x86_64/Darwin/libasyncprocess.so":
  dlopen(/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/static/x86_64/Darwin/libasyncprocess.so, 0x000A): tried: '/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/static/x86_64/Darwin/libasyncprocess.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')).`

**Comment by snmsts** [comment]
Thank you. I can't reproduce what you seeing and seems you don't code in common lisp..

ok... I admit something wrong with my code to determine architecture...

launch sbcl with

```
ros run
```

then put following code which comes from async-process.

```
(eval-when (:compile-toplevel :load-toplevel :execute)
  (defun system (cmd)
    (ignore-errors (string-right-trim '(#\Newline) (uiop:run-program cmd :output :string))))
  (defun muslp ()
    (ignore-errors
      (not (zerop (length (uiop:run-program
                           "ldd /bin/ls |grep musl"
                           :ignore-error-status t
                           :output :string)))))))

(pushnew (asdf:system-relative-pathname
          :async-process
          (format nil "../static/~A/"
                  (cond ((uiop/os:featurep '(:and :windows :x86-64))
                         "x86_64/windows")
                        ((uiop/os:featurep :windows) "x86/windows")
                        ((uiop/os:featurep :unix)
                         (format nil "~A/~A"
                                 (system "uname -m")
                                 (let ((os (system "uname")))
                                   (cond ((and (equal os "Linux")
                                               (muslp))
                                          "Linux-musl")
                                         (t os))))))))
         cffi:*foreign-library-directories*
         :test #'uiop:pathname-equal)
```

then I would like to know the result of 

```
cffi:*foreign-library-directories*
```

also I'd like to confirm ``*feature*``.

```
*features*
```

**Comment by jonathanabennett** [comment]
I'd probably put myself as an advanced beginner in Common Lisp, but it isn't yet my most comfortable programming language (rapidly becoming so though). Here's the result of running `cffi:*foreign-library-directories*`
```(#P"/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/static/x86_64/Darwin/"
 (CFFI::EXPLODE-PATH-ENVIRONMENT-VARIABLE "LD_LIBRARY_PATH")
 (CFFI::EXPLODE-PATH-ENVIRONMENT-VARIABLE "DYLD_LIBRARY_PATH") (UIOP/OS:GETCWD)
 (CFFI::DARWIN-FALLBACK-LIBRARY-PATH))```

and `*features*`:

```(CFFI-FEATURES:FLAT-NAMESPACE CFFI-FEATURES:UNIX CFFI-FEATURES:DARWIN :CFFI
 CFFI-SYS::FLAT-NAMESPACE ALEXANDRIA::SEQUENCE-EMPTYP :QUICKLISP
 :QUICKLISP-SUPPORT-HTTPS :ASDF3.3 :ASDF3.2 :ASDF3.1 :ASDF3 :ASDF2 :ASDF
 :OS-MACOSX :OS-UNIX :NON-BASE-CHARS-EXIST-P :ASDF-UNICODE :ROS.INIT :ARM64
 :GENCGC :64-BIT :ANSI-CL :BSD :COMMON-LISP :DARWIN :IEEE-FLOATING-POINT
 :LITTLE-ENDIAN :MACH-O :PACKAGE-LOCAL-NICKNAMES :SB-CORE-COMPRESSION :SB-LDB
 :SB-PACKAGE-LOCKS :SB-THREAD :SB-UNICODE :SBCL :UNIX)```

I'm wondering if I just completely broke something in setting up the development environment on this, my first M1 Mac. I was planning on reimaging the computer once grades are in for this semester anyway, so if we cannot resolve it, I will try again on a newly imaged computer and see if that addresses it. Though I'd definitely prefer to understand what's going on.

**Comment by snmsts** [comment]
sorry I forgot to load cffi for the snippet. so what I expect to type in repl is 

```
(ql:quickload :cffi)

(eval-when (:compile-toplevel :load-toplevel :execute)
  (defun system (cmd)
    (ignore-errors (string-right-trim '(#\Newline) (uiop:run-program cmd :output :string))))
  (defun muslp ()
    (ignore-errors
      (not (zerop (length (uiop:run-program
                           "ldd /bin/ls |grep musl"
                           :ignore-error-status t
                           :output :string)))))))

(pushnew (asdf:system-relative-pathname
          :async-process
          (format nil "../static/~A/"
                  (cond ((uiop/os:featurep '(:and :windows :x86-64))
                         "x86_64/windows")
                        ((uiop/os:featurep :windows) "x86/windows")
                        ((uiop/os:featurep :unix)
                         (format nil "~A/~A"
                                 (system "uname -m")
                                 (let ((os (system "uname")))
                                   (cond ((and (equal os "Linux")
                                               (muslp))
                                          "Linux-musl")
                                         (t os))))))))
         cffi:*foreign-library-directories*
         :test #'uiop:pathname-equal)
```

and I got

```
(#P"/Users/snmsts/.roswell/local-projects/cxxxr/async-process/static/arm64/Darwin/"
 (CFFI::EXPLODE-PATH-ENVIRONMENT-VARIABLE "LD_LIBRARY_PATH")
 (CFFI::EXPLODE-PATH-ENVIRONMENT-VARIABLE "DYLD_LIBRARY_PATH") (UIOP/OS:GETCWD)
 (CFFI::DARWIN-FALLBACK-LIBRARY-PATH))
```

``async-process``'s path might be different because it is my local repo and I expect you have download it from ``quicklisp``. Never mind difference other than ``arm64/Darwin`` part.

> I'm wondering if I just completely broke something in setting up the development environment on this, my first M1 Mac. I was planning on reimaging the computer once grades are in for this semester anyway, so if we cannot resolve it, I will try again on a newly imaged computer and see if that addresses it. Though I'd definitely prefer to understand what's going on.

I understand. so if you give up on tackling this issue (may be on time up situation?) that'll be ok for me.

My objective for this issue is your sbcl is claiming ``arm64`` (meaning roswell tries ``uname -m`` and got ``arm64``) but my lisp code ``uname -m`` returns ``x86_64``... It's wired and would like to apply fix for ``async-process`` code to support your environment.

**Comment by jonathanabennett** [comment]
Yeah, I want to fix this as well. Our deadline is 10 JUN, that's when I'll reimage this computer and start fresh. Here's what I got on running that. It is getting the x86_64 version. Should I try deleting that directory and trying again?

```* (#P"/Users/jonathan.bennett/.roswell/lisp/quicklisp/dists/quicklisp/software/async-process-20210531-git/static/x86_64/Darwin/"
 (CFFI::EXPLODE-PATH-ENVIRONMENT-VARIABLE "LD_LIBRARY_PATH")
 (CFFI::EXPLODE-PATH-ENVIRONMENT-VARIABLE "DYLD_LIBRARY_PATH") (UIOP/OS:GETCWD)
 (CFFI::DARWIN-FALLBACK-LIBRARY-PATH))
```

**Comment by snmsts** [comment]
mmm

my environment what I'd get is

```
% uname -m
arm64
```

```
 % ros run
* (uiop:run-program "uname -m" :output :string)
"arm64
"
NIL
0
* (quit)
```

I start doubting there's something wrong with your rosetta setup...

**Comment by jonathanabennett** [comment]
Ok, I think I have found what's going on, though now I'm really confused. Here is the output of my shell right now.

```
0 ➜ 2:54PM jonathan.bennett ~ uname -m
x86_64
0 ➜ 2:54PM jonathan.bennett ~ ros roswell-internal-use uname -m
arm64
0 ➜ 2:54PM jonathan.bennett ~ ros run
* (uiop:run-program "uname -m" :output :string)
"x86_64
"
NIL
0
* (quit)
```

**Comment by snmsts** [comment]
did you turn on rosetta?

**Comment by jonathanabennett** [comment]
Yes, I think I remember Docker turning on Rosetta when I installed it a few months ago. And it appears the only way to get rid of it involves a bunch of commands inside System Recovery. Fun.

**Comment by jonathanabennett** [comment]
Ok, because of some school tech administration issues, reimaging my laptop has been moved up to Monday next week anyway. We have identified the problem (Rosetta) and so there's not much more to do. I wish I'd known Rosetta would fubar things up so frustratingly, but that's an issue for me to raise with Apple and I doubt they have the slightest interest in what I think about their design and development choices. I'm going to close this issue as it isn't something that lem can or should fix.
